src/main/scala/org/constellation/snapshotstreaming/storage/FileBasedLastIncrementalCurrencySnapshotStorage.scala

```diff
@@ -0,0 +1,40 @@
+package org.constellation.snapshotstreaming.storage
+
+import cats.effect.std.Semaphore
+import cats.effect.Async
+import cats.syntax.applicative._
+import cats.syntax.applicativeError._
+import cats.syntax.flatMap._
+import cats.syntax.functor._
+import org.tessellation.security._
+import fs2.io.file._
+import io.circe.Codec
+import io.circe.generic.semiauto.deriveCodec
+import org.tessellation.currency.schema.currency.CurrencyIncrementalSnapshot
+import org.tessellation.currency.schema.currency.CurrencySnapshotInfo
+import org.tessellation.currency.schema.currency.CurrencySnapshotStateProof
+import org.tessellation.merkletree.StateProofValidator
+
+class FileBasedLastIncrementalCurrencySnapshotStorage[F[_]: Async: Hasher](semaphore: Semaphore[F], path: Path)(implicit
+  codec: Codec[SnapshotWithState[CurrencyIncrementalSnapshot, CurrencySnapshotInfo]]
+) extends BaseLastIncrementalSnapshotStorage[F, CurrencySnapshotStateProof, CurrencyIncrementalSnapshot, CurrencySnapshotInfo](semaphore, path) {
+
+  def validateStateProof(snapshot: Hashed[CurrencyIncrementalSnapshot], state: CurrencySnapshotInfo): F[Unit] =
+    StateProofValidator
+      .validate(snapshot, state)
+      .map(_.isValid)
+      .flatMap(new Throwable("State proof doesn't match!").raiseError[F, Unit].unlessA)
+
+}
+
+object FileBasedLastIncrementalCurrencySnapshotStorage {
+
+  def make[F[_]: Async: Hasher](path: Path): F[FileBasedLastIncrementalCurrencySnapshotStorage[F]] =
+    Semaphore[F](1).map { semaphore =>
```

Comment:
Yeah, someone used a `Semaphore` with a single permit which is technically a Mutex. I think we can switch to `Mutex`.

Reply:
@sdavidp `Semaphore[F](1)` is a creation of the semaphore. So yes, there is only one fiber that would create a semaphore. but the semaphore itself can be used by multiple threads.

Reply:
@marcinwadon the semaphore(1) (or a mutex) can only be used by other fibers if it's exposed, which this one is not. It _can_ have an effect if multiple fibers use the object returned by `make` but I wanted to point out that in a couple cases that is not happening.

Reply:
I see it's being used by `set` or `setInitial`, so if those methods can be run from multiple places, it can be used from multiple threads, right?

Reply:
@marcinwadon
Case 1: serializes access, only one fiber uses store at a time
```
store <- StorageUsingMutex.make
_ <- fiber_1_sets_storage(store)
_ <- fiber_2_sets_storage(store)
```

Case 2: mutex has no effect because each fiber has its own mutex, so no one else sees/uses it
```
_ <- fiber_1_sets_storage(StorageUsingMutex.make)
_ <- fiber_2_sets_storage(StorageUsingMutex.make)
```

I was pointing out that Case 2 is what's happening in `LastCurrencySnapshotStorage` when using `FileBasedLastIncrementalCurrencySnapshotStorage`

Reply:
@sdavidp I have made the mutex a singleton for `LastCurrencySnapshotStorage` and I am just passing it to temp storages. So only one currency snapshot can be proceeded at once.

Reply:
> So only one currency snapshot can be proceeded at once.

@kpudlik I'm not certain this is true. Isn't my Case 2 an example of what we're doing? If so, do you agree that the mutex has no effect for _external calls_ to `set` ?  Or is the intention to serialize the stream _inside_ `set` ?

Reply:
@kpudlik I did another pull and now I see your changes. Yes, for `LastCurrencySnapshotStorage.set` since the mutex is now passed into `FileBasedLastIncrementalCurrencySnapshotStorage.make` that will work. 👍 

Just in case it's important, that's not happening for `getLastBalances` -- it's still using its own mutex so no mutual exclusion there.